lib/AST/ClangTypeConverter.cpp

```diff
@@ -883,6 +900,93 @@ Decl *ClangTypeConverter::getSwiftDeclForExportedClangDecl(
   return (it != ReversedExportMap.end() ? it->second : nullptr);
 }
 
+clang::QualType ClangTypeConverter::convertTemplateArgument(Type type) {
+  // C++ function templates can only be instantiated with Clang types and
+  // a handful of Swift builtin types. These are enumerated here rather than
+  // delegated to ClangTypeConverter::convert() (which is more general).
+  auto withCache = [&](auto lookup) {
+    auto cached = Cache.find(type);
+    if (cached != Cache.end())
+      return cached->second;
+
+    auto result = lookup();
+    if (!result.isNull())
+      Cache.insert({type, result});
```

Comment:
I think it is possible to replace the double lookup in `Cache` case of a cache miss using `try_emplace`. Something like:
```
auto [it, inserted] = Cache.try_emplace(type, clang::QualType{});
if (!inserted)
  it->second = lookup();
return it->second;
```

Also, if `lookup()` returns null, do we want to repeat it the next call? Is there any reason to expect a different result next time?

Reply:
I see there are other parts of this file that are doing double lookups in case of cache misses, up to you if you want to fix those as well.

Reply:
> Also, if lookup() returns null, do we want to repeat it the next call? Is there any reason to expect a different result next time?

Probably not, but I'm not fully confident of that. Do you reckon it's worth inserting an `#ifdef NDEBUG` to catch such circumstances?

Reply:
> I see there are other parts of this file that are doing double lookups in case of cache misses, up to you if you want to fix those as well.

I think it might be better to fix those in a follow-up patch, since this one is starting to get large. I'll change my use of `Cache.find()` though.

Reply:
Er, sanity check: you meant `if (inserted)`, right?

```cpp
auto [it, inserted] = Cache.try_emplace(type, clang::QualType{});
if (inserted)
  it->second = lookup();
return it->second;
```

`inserted == true` would imply there was previously no entry, which is when we would need to perform a `lookup()`. If `inserted == false` then there was already a cached entry.

Reply:
Sorry, my bad! You are right!

Reply:
As it turns out, `DenseMap::try_emplace()` does not work because the returned iterator may get invalidated by insertions during nested conversions, i.e.:

```cpp
auto [it, inserted] = Cache.try_emplace(type, clang::QualType{});
if (inserted)
  it->second = lookup(); // lookup() may cause insertions that invalidate it
return it->second;
```

So I'm going to revert that change.

(Also I realize that `lookup()` is not the best choice of variable name; I've renamed it to `conversion()`.)

Reply:
Also, following up on this:

> Also, if lookup() returns null, do we want to repeat it the next call? Is there any reason to expect a different result next time?

It turns out that it _is_ necessary, even though the original `convert()` function did not appear to do this. I don't remember why I even added that check, or how the old `convert()` function was working before... I'll leave it as is for now but investigate that in a follow-up patch.